### ISSUE TITLE
amesos2 klu2_internal.h defining NDEBUG

### ISSUE THREAD

**Issue by crdrumm** [trilinos/Trilinos#3978]
@trilinos/amesos2

## Current Behavior

In klu2_internal.h the following lines define NDEBUG which disables assertions in SCEPTRE unit tests

/* make sure debugging and printing is turned off */

#ifndef NDEBUG
#define NDEBUG
#endif

/* To enable debugging and assertions, uncomment this line:
 #undef NDEBUG
 */

To be more specific, we have the following line in our code:

#include "Amesos2.hpp"

That file has the following chain of #includes:
#include "Amesos2_Factory.hpp"
#include "Amesos2_KLU2.hpp"
#include "Amesos2_KLU2_decl.hpp"
#include "Amesos2_KLU2_FunctionMap.hpp"
#include "klu2_defaults.hpp"
#include "klu2_internal.h"

In particular, the file Amesos2_KLU2_FunctionMap.hpp has the following section:

/* External definitions of the KLU2 functions
 *
 */
 // TODO
namespace KLU2 {
#include "klu2_defaults.hpp"
#include "klu2_analyze.hpp"
#include "klu2_factor.hpp"
#include "klu2_solve.hpp"
#include "klu2_tsolve.hpp"
#include "klu2_free_symbolic.hpp"
#include "klu2_free_numeric.hpp"
} // end namespace KLU2

The namespace does nothing to protect against preprocessor commands, which I suspect was the intention of the KLU2 namespace.

## Motivation and Context
assertions have been unexpectedly disabled, so tests compiled with -g erroneously passed

## Your Environment

version 12.12.1

  -D Trilinos_VERBOSE_CONFIGURE:BOOL=ON \
  -D CMAKE_VERBOSE_MAKEFILE:BOOL=TRUE \
  -D CMAKE_BUILD_TYPE:STRING=RELEASE \
  -D CMAKE_CXX_FLAGS="-O3" \
  -D TPL_ENABLE_BinUtils:BOOL=OFF \
  -D TPL_ENABLE_MPI:BOOL=ON \
  -D MPI_BASE_DIR:PATH=$MPIHOME \
  -D TPL_ENABLE_BLAS:BOOL=ON \
  -D TPL_ENABLE_LAPACK:BOOL=ON \
  -D LAPACK_LIBRARY_NAMES:STRING="mkl_core;mkl_sequential;mkl_intel_lp64" \
  -D LAPACK_LIBRARY_DIRS:PATH=$MKL_LIB \
  -D HAVE_TEUCHOS_LAPACKLARND=OFF \
  -D BLAS_LIBRARY_NAMES:STRING="mkl_core;mkl_intel_lp64;mkl_intel_thread;iomp5;dl" \
  -D BLAS_LIBRARY_DIRS:PATH="$MKL_LIB;$INTEL_LIB" \
  -D TPL_ENABLE_Boost:BOOL=ON \
  -D Boost_INCLUDE_DIRS:PATH=$CODE_PATH/boost/1.67.0/boost_1_67_0 \
  -D TPL_ENABLE_Netcdf:BOOL=ON \
  -D Netcdf_LIBRARY_DIRS:PATH=$CODE_PATH/netcdf/4.6.1/$COMPILER_VERS\_install/lib \
  -D Netcdf_INCLUDE_DIRS:PATH=$CODE_PATH/netcdf/4.6.1/$COMPILER_VERS\_install/include \
  -D TPL_ENABLE_Matio:BOOL=OFF \
  -D TPL_ENABLE_X11:BOOL=OFF \
  -D Trilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \
  -D Trilinos_ENABLE_Amesos2:BOOL=ON \
  -D Amesos2_ENABLE_Basker:BOOL=ON \
  -D HAVE_AMESOS2_LAPACK:INTERNAL=ON \

gcc 4.9.3
openmpi 1.10

boost 1.67.0
netcdf 4.6.1

- **Relevant repo SHA1s:**  
- **Relevant configure flags or configure script:**  
- **Operating system and version:**  
- **Compiler and TPL versions:**  

## Related Issues

* Blocks 
* Is blocked by 
* Follows 
* Precedes 
* Related to 
* Part of 
* Composed of

**Comment by mhoemmen** [comment]
I don't think Clif's tagging took (I think only Trilinos developers can tag groups): @trilinos/amesos2

**Comment by ndellingwood** [comment]
PR #4528 issued that should fix this.

**Comment by ndellingwood** [comment]
PR #4528 with fix merged into develop branch, closing the issue.
